engine/src/render/cascaded_shadow_maps_rasterizer/plugin.cpp

```diff
@@ -0,0 +1,314 @@
+                        {
+                            corner = view * corner;
+                        }
+                        // Find minimum/maximum coordinates along each axis
+                        float minX = std::numeric_limits<float>::max();
+                        float maxX = std::numeric_limits<float>::lowest();
+                        float minY = std::numeric_limits<float>::max();
+                        float maxY = std::numeric_limits<float>::lowest();
+                        float minZ = std::numeric_limits<float>::max();
+                        float maxZ = std::numeric_limits<float>::lowest();
+                        for (const auto& corner : cameraFrustumCorners)
+                        {
+                            minX = std::min(minX, corner.x);
+                            maxX = std::max(maxX, corner.x);
+                            minY = std::min(minY, corner.y);
+                            maxY = std::max(maxY, corner.y);
+                            minZ = std::min(minZ, corner.z);
+                            maxZ = std::max(maxZ, corner.z);
+                        }
+
+                        // Expand space between Z planes, so that objects outside the frustum can cast shadows
+                        minZ -= std::abs(minZ) * 0.5F;
+                        maxZ += std::abs(maxZ) * 0.5F;
+                        auto proj = glm::ortho(minX, maxX, minY, maxY, -maxZ, -minZ);
+                        perScene.lightViewProj[i] = proj * view;
+                        perScene.numSplits = caster.numSplits;
+                    }
+                    state.perSceneCB->fill(&perScene, sizeof(perScene));
+
+                    // Bind the shader, vertex array and uniform buffer.
+                    rd.setShaderPipeline(state.pipeline);
+                    rd.setVertexArray(state.vertexArray);
+                    state.perSceneBP->bind(state.perSceneCB);
+                    state.perMeshBP->bind(state.perMeshCB);
+
+                    // Iterate over all mesh buckets and issue draw calls.
+                    for (auto [meshEnt, meshLocalToWorld, mesh, grid] : meshes)
+                    {
+                        // Send the PerMesh data to the GPU.
+                        PerMesh perMesh{
+                            .model = meshLocalToWorld.mat * glm::translate(glm::mat4(1.0F), grid.offset),
+                        };
+                        state.perMeshCB->fill(&perMesh, sizeof(perMesh));
+
+                        // Iterate over the buckets of the mesh (it may be split over many of them).
+                        for (auto bucket = mesh.firstBucketId; bucket != RenderMeshPool::BucketId::Invalid;
+                             bucket = pool.next(bucket))
+                        {
+                            rd.drawTriangles(pool.bucketSize() * bucket.inner, pool.vertexCount(bucket));
+                        }
+                    }
```

Comment:
Not something to fix right now, but worth pointing out I guess.
The fact that we have this logic duplicated over at least three files as of now smells of code smell to me.
We should think about this further; will open an issue.